packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts

```diff
@@ -384,23 +388,69 @@ export const useGridSelection = (
 
   const handleCellKeyDown = React.useCallback<GridEventListener<GridEvents.cellKeyDown>>(
     (params, event) => {
+      // Get the most recent cell mode because it may have been changed by another listener
+      if (apiRef.current.getCellMode(params.id, params.field) === GridCellModes.Edit) {
+        return;
+      }
+
+      if (isNavigationKey(event.key) && event.shiftKey) {
+        const focusCell = gridFocusCellSelector(apiRef);
+        if (focusCell && focusCell.id !== params.id) {
+          event.preventDefault();
+
+          const isPreviousRowSelected = apiRef.current.isRowSelected(focusCell.id);
+          if (canHaveMultipleSelection) {
+            const newRowIndex = apiRef.current.getRowIndexRelativeToVisibleRows(focusCell.id);
+            const previousRowIndex = apiRef.current.getRowIndexRelativeToVisibleRows(params.id);
+
+            let start: number;
+            let end: number;
+
+            if (newRowIndex > previousRowIndex) {
+              if (isPreviousRowSelected) {
+                // We are navigating to the bottom of the page and adding selected rows
+                start = previousRowIndex;
+                end = newRowIndex - 1;
+              } else {
+                // We are navigating to the bottom of the page and removing selected rows
+                start = previousRowIndex;
+                end = newRowIndex;
+              }
+            } else {
+              // eslint-disable-next-line no-lonely-if
+              if (isPreviousRowSelected) {
+                // We are navigating to the top of the page and removing selected rows
+                start = newRowIndex + 1;
+                end = previousRowIndex;
+              } else {
+                // We are navigating to the top of the page and adding selected rows
+                start = newRowIndex;
+                end = previousRowIndex - 1;
+              }
+            }
+
+            const rowsBetweenStartAndEnd = visibleRows.rows
+              .slice(start, end + 1)
+              .map((row) => row.id);
+            apiRef.current.selectRows(rowsBetweenStartAndEnd, !isPreviousRowSelected);
+          } else {
+            apiRef.current.selectRow(focusCell.id, !isPreviousRowSelected, true);
```

Comment:
Could you invert `canHaveMultipleSelection` -> `!canHaveMultipleSelection` and put this logic at the top, with a immediate return? The logic above is a bit long so we scroll a few lines to see only 1 in the else block. The sooner we return, the better to read.